mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,234 @@
+                    )
+                    if self.save_sampled_queries:
+                        self.output_manager.update_selected_queries_json(
+                            strategy,
+                            epoch,
+                            iteration,
+                            self.data_bucket.newly_sampled_queries,
+                        )
+                    if self.early_stopping_window > 0:
+                        early_stop = self.output_manager.check_early_stopping(
+                            strategy, self.early_stopping_window
+                        )
+                num_unsampled = len(self.data_bucket.unsampled_queries)
+                if num_unsampled > 0:
+                    sampling_size = (
+                        self.batch_size
+                        if num_unsampled > self.batch_size
+                        else num_unsampled
+                    )
+                    heuristic = HeuristicsFactory.get_heuristic(strategy, sampling_size)
+                    (
+                        self.data_bucket.newly_sampled_queries,
+                        self.data_bucket.sampled_queries,
+                        self.data_bucket.unsampled_queries,
+                        _,
+                    ) = heuristic.sample(
+                        sampled=self.data_bucket.sampled_queries,
+                        unsampled=self.data_bucket.unsampled_queries,
+                        preds_single=preds_single,
+                        preds_multi=preds_multi,
+                        domain_indices=domain_indices,
+                    )
+                if selection_mode or early_stop:
+                    return
+                iteration += 1
+
+    def select_queries_to_label(self):
+        """Selects the next batch of queries to label from a set of log queries."""
+        logger.info("Loading Queries for Active Learning.")
+        self.init_data_bucket = DataBucketFactory.get_data_bucket_for_selection(
+            self.app_path,
+            self.load,
+            self.save,
+            self.train_pattern,
+            self.test_pattern,
+            self.unlabeled_logs_path,
+            self.labeled_logs_pattern,
+            self.log_usage_pct,
+        )
+        logger.info("Starting Selection.")
+        self._train_strategy(strategy=self.log_selection_strategy, selection_mode=True)
+        self.output_manager.write_log_selected_queries_json(
+            strategy=self.log_selection_strategy,
+            queries=self.data_bucket.newly_sampled_queries,
+        )
+
+    def plot(self):
+        """Creates the generated folder and its subfolders if they do not already exist."""
+        plot_manager = PlotManager(self.output_manager.experiment_folder)
+        plot_manager.generate_plots()
+
+
+# TODO: This function is temporary. Replace with better data validation
+def flatten_active_learning_config(original_config):
+    """Create a flattened config to use as params.
```

Comment:
I am not very clear why we have this function. Why can't we just access the relevant sub-keys for whatever access pattern that is desired?

Reply:
Removed in favor of an ActiveLearningPipelineFactory class that passes in sub-keys as suggested.